Add driver support for Stellaris ethernet controller.

bravegnu opened the pull request.

bravegnu committed:
net: ethernet: Add helper functions to update error stats.

Add missed out helper functions to update the errors.tx and error.rx
in the stats structure.

Signed-off-by: Fadhel Habeeb <fadhel@zilogic.com>
Signed-off-by: Nirav Parmar <niravparmar@zilogic.com>
Signed-off-by: Vijay Kumar B <vijaykumar@zilogic.com>

tbursztyka reviewed (dismissed):
Please, could you split your 2nd patch in 3 parts: 
- driver addition
- dts binding part
- samples part

bravegnu committed:
drivers: ethernet: Add TI Stellaris ethernet controller driver.

The driver can be tested using different networking emulation
approaches.

This approach will work across multiple Qemu instances. There can be
more than one Qemu instance, run using the following command. They
would appear to be on the same Ethernet network.

  $ qemu-system-arm -M lm3s6965evb                      \
                    -serial stdio                       \
                    -net nic                            \
                    -net socket,mcast=230.0.0.1:1234    \
                    -kernel zephyr.elf

This approach will work with other virtualization technologies that
support connecting to a VDE switch, like VirtualBox and User Mode
Linux. The switch can be started using the following command.

  $ vde_switch --sock /tmp/switch

Qemu can be connected to the switch using the following command.

  $ qemu-system-arm -M lm3s6965evb                      \
                    -serial stdio                       \
                    -net nic                            \
                    -net vde,sock=/tmp/switch   	\
                    -kernel zephyr.elf

Signed-off-by: Fadhel Habeeb <fadhel@zilogic.com>
Signed-off-by: Nirav Parmar <niravparmar@zilogic.com>
Signed-off-by: Vijay Kumar B <vijaykumar@zilogic.com>

bravegnu committed:
drivers: ethernet: Add DTS support for Stellaris Ethernet controller.

  * Add DTS support for Stellaris ethernet controller.
  * Add base DTS binding definition for Ethernet.
  * Add DTS binding definition for Stellaris ethernet controller.

Signed-off-by: Fadhel Habeeb <fadhel@zilogic.com>
Signed-off-by: Nirav Parmar <niravparmar@zilogic.com>
Signed-off-by: Vijay Kumar B <vijaykumar@zilogic.com>

bravegnu committed:
samples: net: Modify echo server / client for Stellaris Ethernet.

Add project configuration files for echo_server & echo_client.

Signed-off-by: Fadhel Habeeb <fadhel@zilogic.com>
Signed-off-by: Nirav Parmar <niravparmar@zilogic.com>
Signed-off-by: Vijay Kumar B <vijaykumar@zilogic.com>

bravegnu commented:
@tbursztyka The patch has been split as requested.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+	sys_write32(isr_val, REG_MACRIS);
+
+	if (isr_val & BIT_MACRIS_RXINT) {
+
```
Extra empty line can be removed.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	if (isr_val & BIT_MACRIS_TXEMP) {
+
```
ditto

bravegnu replied:
Empty lines have been removed.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	if (isr_val & BIT_MACRIS_TXER) {
+
```
here too

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	if (isr_val & BIT_MACRIS_RXER) {
+
```
and here

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+
+		SYS_LOG_ERR("Error Frame Recieved");
+		eth_stellaris_rx_error(dev_data->iface);
+	}
```
please place empty line after }

bravegnu replied:
This has been added.

jukkar commented on samples/net/echo_client/boards/qemu_cortex_m3.conf:
```diff
@@ -0,0 +1,10 @@
+# Select Stellaris Ethernet
```
So after this it is not possible to use TAP device to connect to host system, is this really desired?

bravegnu replied:
Qemu's user mode networking can be used to connect to the host system. This is much simpler than SLIP+TAP. Also Qemu supports TAP, VDE, and socket backends as well.

jukkar replied:
> Qemu's user mode networking can be used to connect to the host system.

You just disabled TAP with this config by default. To me it looks like you should place the Stellaris setup to an overlay config file. User can select that then if needed when compiling the application.

jukkar commented on samples/net/echo_client/boards/qemu_cortex_m3.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::2"
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::1"
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.2"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.1"
```
Add `<enter>` at the end of line.

bravegnu replied:
This has been added, as well.

jukkar commented on samples/net/echo_server/boards/qemu_cortex_m3.conf:
```diff
@@ -0,0 +1,13 @@
+
```
Same here, TAP connectivity no longer posible with qemu_cortex_m3

bravegnu replied:
Retaining the original sample, and added separate overlay.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,366 @@
+};
+
+struct eth_stellaris_runtime eth_data = {
+	.mac_addr = {
```
You are allowing user to specify any MAC address here, should we limit this to bytes 3-6 as the first three should be TI specific?

bravegnu replied:
We are not very sure about this, there doesn't seem to be any specific OUI recommended by TI.

jukkar replied:
We are using 00-12-4B in cc2520 and cc1200 drivers, I suppose that could be used here too. There are also other options mentioned in https://hwaddress.com/?q=texas%20instruments

bravegnu replied:
@jukkar Not sure why we would want to do that. The user should be free to assign any MAC address to the interface.

jukkar replied:
User can certainly change anything as the code can be tweaked. I was just suggesting some sane defaults for the default mac address.

bravegnu replied:
The DTS has been updated to use the RFC 7042 MAC address that you suggested as the default.

galak reviewed (dismissed):
some dt feedback

bravegnu commented on dts/bindings/ethernet/ti,stellaris-ethernet.yaml:
```diff
@@ -0,0 +1,28 @@
+    This binding gives a base representation of the TI Stellaris Ethernet
+
+inherits:
+  !include ethernet.yaml
```
@galak The file is in dts/bindings/ethernet/ethernet.yaml

bravegnu replied:
@galak Yes, it is available in the commit c1f4221216c7303fad9459eb99d5ce5d42792c67.

jukkar commented on dts/arm/ti/lm3s6965.dtsi:
```diff
@@ -46,6 +46,15 @@
+			reg = <0x40048000 0x3c>;
+			interrupts = <42 0>;
+			status = "disabled";
+			local-mac-address = [00 11 22 33 44 55];
```
If you set some default mac address, then at least use the range allocated for documentation 00-00-5E-00-53-xx, see  RFC 7042 for details.

bravegnu replied:
Updated MAC as per your suggestion.

jukkar commented on samples/net/echo_client/boards/qemu_cortext_m3.conf:
```diff
@@ -1,7 +0,0 @@
-# The addresses are selected so that qemu<->qemu connectivity works ok.
```
Note sure why this was removed. Just keep this file, and put the stellaris specific stuff into overlay-stellaris.conf file.

jukkar commented on samples/net/echo_client/overlay-qemu_cortex_m3_eth.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::2"
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::1"
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.2"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.1"
```
The line is missing \n

bravegnu replied:
This has been fixed.

jukkar commented on samples/net/echo_server/overlay-qemu_cortex_m3_eth.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1"
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::2"
```
same here

bravegnu replied:
Fixed this as well.

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+	u8_t *eth_hdr;
+	struct net_eth_hdr *pkt_hdr;
+	struct device *dev = net_if_get_device(iface);
+	struct eth_stellaris_runtime *dev_data = DEV_DATA(dev);
```
Only one comment on that driver now, do as you did everywhere else (and as we do ourselves also): pre-assigned variables come first in the list of declaration.

bravegnu replied:
This has been fixed.

pfalcon commented:
@bravegnu: Great work, but the same question as for https://github.com/zephyrproject-rtos/zephyr/pull/9112 : how to use this?

pfalcon commented:
Also currently has checkpatch issues (CI is red):

~~~
-:189: WARNING:TYPO_SPELLING: 'successfull' may be misspelled - perhaps 'successful'?
#189: FILE: drivers/ethernet/eth_stellaris.c:127:
+	/* Wait and check if txn successfull or not */

-:349: WARNING:TYPO_SPELLING: 'Recieved' may be misspelled - perhaps 'Received'?
#349: FILE: drivers/ethernet/eth_stellaris.c:287:
+		SYS_LOG_ERR("Error Frame Recieved");
~~~

etc.

pfalcon commented on drivers/ethernet/Kconfig.stellaris:
```diff
@@ -0,0 +1,14 @@
+menuconfig ETH_STELLARIS
+	bool "Enable TI Stellaris MCU family ethernet driver."
+	depends on NET_L2_ETHERNET
+	default n
```
No need for "default n".

bravegnu replied:
This has been removed.

pfalcon commented:
> how to use this?

Answering my own question:

1. Build echo_server with overlay-qemu_cortex_m3_eth.conf as included in this patch.
2. Run: (not sudo, required for qemu to create TAP device):

~~~
sudo /home/pfalcon/opt/zephyr-sdk-0.9.3/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-arm \
-cpu cortex-m3 -machine lm3s6965evb -nographic -vga none \
-pidfile qemu.pid \
-serial   mon:stdio \
-netdev tap,id=netdev1 \
-net nic,model=stellaris,netdev=netdev1 \
-object filter-dump,id=filter1,netdev=netdev1,file=dump-m3.dat \
-kernel /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/samples/net/echo_server/outdir/qemu_cortex_m3/zephyr/zephyr.elf
~~~

3. `ifconfig tap0` 192.0.2.2 as usual
4. `ping -I tap0 192.0.2.1`. Pings might not work at once (same random mac/arp problem?), might not work at all. But mostly they do.

Seen on Zephry side:
~~~
stellaris_enet: error: RX underflow
stellaris_enet: error: RX underflow
stellaris_enet: error: RX underflow
stellaris_enet: error: RX underflow
~~~

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "dev/eth_stellaris"
```
While this was fixed, the syslog is being deprecated, so these two syslog settings need to be replaced by
```
#define LOG_MODULE_NAME eth_stellaris
#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
#include <logging/log.h>
LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```

bravegnu replied:
This has been fixed.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+#define SYS_LOG_DOMAIN "dev/eth_stellaris"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_ETHERNET_LEVEL
+
+#include <logging/sys_log.h>
```
This sys_log.h can be removed

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+	struct net_eth_hdr *pkt_hdr;
+
+	if (!pkt->frags) {
+		SYS_LOG_ERR("No data to send");
```
Search and replace `s/SYS_//` in this file.

bravegnu replied:
Done.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+	/* Update statistics counters */
+	eth_stats_update_bytes_tx(iface, net_pkt_get_len(pkt));
+	eth_stats_update_pkts_tx(iface);
+	pkt_hdr = NET_ETH_HDR(pkt);
```
I think you could remove the pkt_hdr variable. If CONFIG_NET_STATISTICS_ETHERNET is not enabled, then having this extra variable is useless here. Just use NET_ETH_HDR(pkt) directly.

Or we could write it like this too in which case we could use pkt_hdr
```
if (IS_ENABLED(CONFIG_NET_STATISTICS_ETHERNET)) {
        struct net_eth_hdr *pkt_hdr;

	/* Update statistics counters */
	eth_stats_update_bytes_tx(iface, net_pkt_get_len(pkt));
	eth_stats_update_pkts_tx(iface);
	pkt_hdr = NET_ETH_HDR(pkt);

	if (net_eth_is_addr_multicast(&pkt_hdr->dst)) {
		eth_stats_update_multicast_tx(iface);
	} else if (net_eth_is_addr_broadcast(&pkt_hdr->dst)) {
		eth_stats_update_broadcast_tx(iface);
	}
}
```

Same applies to receive function.

bravegnu replied:
Going with the second approach.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+	struct device *dev = net_if_get_device(iface);
+
+	eth_stats_update_errors_rx(iface);
+	/* Clear the rx_frame buffer,
```
Can you place empty line before the comment, it would look better that way.

bravegnu replied:
Done

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+			bytes_left = 0;
+			break;
+		}
+		/* Read the the partial word */
```
Empty line before comment please.

bravegnu replied:
Done.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+				    (u8_t *)&reg_val, K_NO_WAIT)) {
+			return -1;
+		}
+		bytes_left -= 4;
```
Place empty lines after } as that is the style in zephyr.

bravegnu replied:
Done.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+		}
+		bytes_left -= 4;
+	}
+	return frame_len;
```
same here

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	/* Update statistics counters */
+	eth_stats_update_bytes_rx(iface, frame_len - 6);
```
Same comment as in tx func

bravegnu replied:
Done.

jukkar commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,365 @@
+		&eth_stellaris_apis, ETHERNET_L2,
+		NET_L2_GET_CTX_TYPE(ETHERNET_L2), ETH_MTU);
+
+
```
Remove extra empty line

jukkar commented on samples/net/echo_client/overlay-qemu_cortex_m3_eth.conf:
```diff
@@ -0,0 +1,11 @@
+# Select Stellaris Ethernet
```
Is there reason why we should not call this overlay-stellaris.conf?

bravegnu replied:
The other overlay files seem to following the naming convention overlay-board-device.conf, when the overlay file is specific to board.

jukkar replied:
Ok, this makes sense.

jukkar commented:
ping @bravegnu any progress with this one, it looks like with minor tweaks this could be merged soon.

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,378 @@
+	reg_val = sys_read32(REG_MACDATA);
+	frame_len = reg_val & 0x0000ffff;
+
+	if (!net_pkt_append(pkt, 2, (u8_t *)&reg_val + 2, K_NO_WAIT)) {
```
could you add a comment explaining what it is doing (I guess the 2 bytes are the start of the actual ethernet frame)

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,378 @@
+
+	frame_len = eth_stellaris_rx_pkt(dev, pkt);
+	if (frame_len == -1) {
+		goto err_rx;
```
move the LOG_ERR here and use pkt_unref as the goto

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,378 @@
+	struct eth_stellaris_config *dev_conf;
+	struct eth_stellaris_runtime *dev_data;
+
+	dev = net_if_get_device(iface);
```
you could already do that in above struct device dev declaration actually.

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,378 @@
+		&eth_stellaris_apis, ETHERNET_L2,
+		NET_L2_GET_CTX_TYPE(ETHERNET_L2), ETH_MTU);
+
+static void eth_stellaris_irq_config(struct device *dev)
```
move that above your eth_cfg struct

bravegnu replied:
This is not possible, because eth_stellaris_irq_config() references the eth_stellaris device structure. The device structure is defined by NET_DEVICE_INIT(), and hence the eth_stellaris_irq_config can only appear after NET_DEVICE_INIT().

bravegnu replied:
@pfalcon Thanks for the tip, this has been fixed.

tbursztyka replied:
stm32 is especially complex, but yes that's the way.

I am asking that we stay consistent because when it will be time to move these macros to codegen, it will save quite some time.

tbursztyka reviewed (changes requested):
minor comments.
Revise where you put init/irq_config and NET_DEVICE_INIT stuff. See other drivers, basically the NET_DEVICE_INIT macro is always the last.

pfalcon commented:
@bravegnu, As requests for fixups keep coming in, can you please also add info on how to actually run it to one of commit messages? I may suggest to the commit "drivers: ethernet: Add TI Stellaris ethernet controller driver.", why now has just repetition of "* Add Stellaris ethernet controller driver."

My research on how to get it run is in https://github.com/zephyrproject-rtos/zephyr/pull/10097#issuecomment-428745209, but as the author, you surely would no better how to run it.

Thanks in advance.

tbursztyka commented on drivers/ethernet/eth_stellaris.c:
```diff
@@ -0,0 +1,399 @@
+	return 0;
+}
+
+static struct device DEVICE_NAME_GET(eth_stellaris);
```
You should not need this

bravegnu replied:
But didn't we agree to this: https://github.com/zephyrproject-rtos/zephyr/pull/10097#discussion_r230334164

tbursztyka replied:
Ah yes, my bad. I am used to other type of drivers (spi, gpio...) and there the irq config function comes after the DEVICE_INIT() macro call. Nevermind.

pfalcon reviewed (approved):
@bravegnu, Thanks for all the changes!

pfalcon commented:
@bravegnu: Here's spur-of-the-moment reply to your concerns: https://github.com/zephyrproject-rtos/zephyr/pull/10547#issuecomment-437357022

You can read it yourself and assess, what I concentrate on is phrase "Like if the ethernet device would be dead. And the same behaviour happens on master too".

And that's only "happens just as we speak" activity. We have many other corner cases yet which we need to investigate and fix. The usual problem is that it happens only to one party and is not/cannot be reproduced by others.

pfalcon commented:
@bravegnu:

> I would like to bring to your notice the following issue, occurs when testing the Stellaris ethernet driver.

Or look at the results of another qemu eth driver: https://github.com/zephyrproject-rtos/zephyr/pull/11218#issuecomment-437926004

That's why we need qemu eth drivers - to be able to reproduce these issues consistently and then bust them.

pfalcon commented:
> Could you try if using stellaris ethernet driver works seamless too with those kind of changes?

IMHO, that can be part of a separate PR, we can't creep the scope of PRs (especially of first-time contributors) indefinitely. I vote for this ro merged as is.

@galak : Can you please re-review?

jukkar commented:
> IMHO, that can be part of a separate PR, we can't creep the scope of PRs (especially of first-time contributors) indefinitely. I vote for this ro merged as is.

Sure, I was not really suggesting that this PR would contain that stuff I mentioned in my comment. It was just a hint that stellaris stuff might work similar way as e1000 with quite minimal changes.

bravegnu commented:
@galak @MaureenHelm :bell: A gentle reminder.

pfalcon commented:
> That PR introduced a way to seamlessly use the ethernet driver in Qemu if configured properly. Part of this configuration is to set the nic model to be used in Qemu command line. See for example how e1000 is doing it in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ethernet/Kconfig.e1000
Could you try if using stellaris ethernet driver works seamless too with those kind of changes?

@bravegnu, Friendly ping to see if the above is still in your queue.

pfalcon commented:
>> That PR introduced a way to seamlessly use the ethernet driver in Qemu if configured properly. ...

I prepared a patch for that, https://github.com/zephyrproject-rtos/zephyr/pull/12422
